File: 프로그래머스/0/181916. 주사위 게임 3/주사위 게임 3.js
function solution(a, b, c, d) {
  const counts = {};
  for (const n of [a, b, c, d]) {
    counts[n] = (counts[n] || 0) + 1;
  }

  const size = new Set([a, b, c, d]).size;
  const diceVals = Object.values(counts);
  const diceKeys = Object.keys(counts).map(Number);

  if (size === 1) {
    return 1111 * a;
  }
  if (size === 4) {
    return Math.min(a, b, c, d);
  }
  if (size === 2 && diceVals.includes(3)) {
    const p = diceKeys.find((x) => counts[x] === 3);
    const q = diceKeys.find((x) => counts[x] === 1);
    return (10 * p + q) ** 2;
  }
  if (size === 2) {
    const [p, q] = diceKeys.filter((x) => counts[x] === 2);
    return (p + q) * Math.abs(p - q);
  }

  if (size === 3) {
    const p = diceKeys.find((x) => counts[x] === 2);
    const [q, r] = diceKeys.filter((x) => counts[x] === 1);
    return q * r;
  }
}

File: 프로그래머스/0/181916. 주사위 게임 3/주사위 게임 3.py
def solution(a, b, c, d):
    counts = {}
    for n in [a, b, c, d]:
        counts[n] = counts.get(n, 0) + 1

    size = len(set([a, b, c, d]))

    if size == 1:
        return 1111 * a
    if size == 4:
        return min(a, b, c, d)
    if size == 2 and 3 in counts.values():
        p = next(k for k, v in counts.items() if v == 3)
        q = next(k for k, v in counts.items() if v == 1)
        return (10 * p + q) ** 2
    if size == 2:
        p, q = [k for k, v in counts.items() if v == 2]
        return (p + q) * abs(p - q)
    if size == 3:
        q, r = [k for k, v in counts.items() if v == 1]
        return q * r